core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala

```diff
@@ -51,18 +58,25 @@ private[kafka] abstract class SubSourceLogic[K, V, Msg](
 
     subscription match {
       case TopicSubscription(topics) =>
-        consumer.tell(KafkaConsumerActor.Internal.Subscribe(topics, rebalanceListener), self.ref)
+        consumer.tell(KafkaConsumerActor.Internal.Subscribe(topics, rebalanceListener, loadOffsetsOnAssign.isEmpty), self.ref)
       case TopicSubscriptionPattern(topics) =>
-        consumer.tell(KafkaConsumerActor.Internal.SubscribePattern(topics, rebalanceListener), self.ref)
+        consumer.tell(KafkaConsumerActor.Internal.SubscribePattern(topics, rebalanceListener, loadOffsetsOnAssign.isEmpty), self.ref)
     }
   }
 
+  implicit val seekTimeout = Timeout(5000, TimeUnit.MILLISECONDS) // TODO: configurable? retry?
   val partitionAssignedCB = getAsyncCallback[Iterable[TopicPartition]] { tps =>
     pendingPartitions ++= tps.filter(!partitionsInStartup.contains(_))
-    pump()
+    loadOffsetsOnAssign.foreach { loadOffsets =>
+      implicit val ec = materializer.executionContext
+      loadOffsets(tps.toSet)
+        .flatMap { offsets => consumer.ask(KafkaConsumerActor.Internal.Seek(offsets)) }
+        .foreach(_ => pump())
```

Comment:
This seems dangerous to me since `pump` does some mutation, but it was already being called in an asynchronous callback anyway so maybe it's okay? If the `seekTimeout` above is shorter than the `group.initial.rebalance.delay.ms` config value on the brokers, then it should never be a problem.

Reply:
yes, you can't call `pump()` from the outside. Previously it was from the inside because it was in the async callback, but now it's `future.foreach` which could be running in another thread (i.e. outside). You must call `pump()` via another async callback.

Reply:
what happens when `loadOffsetsOnAssign` is `None`? Previously there was a call to `pump()` but now that is gone?

Reply:
Ya you're right I still need to call `pump()` when it's `None`. I guess I don't understand how the async callback works. To me it sounds like that would be running in another thread, but it's not?

Reply:
Okay I think I get it now. So if I have `val pumpCB = getAsyncCallback[Unit](_ => pump())`, can I do:
```
...
.foreach { _ => pumpCB.invoke()) }
```
?

Reply:
yes, that's right

you can think of the stage as an actor, onPush, onPull are like messages and are running inside the stage/actor. Same with the async callback functions, they are running inside and therefore safe to update mutable state inside the stage. The purpose is that when you run outside the stage, such as the future in this case, you want "send" that to the stage via callback.invoke.